Lib/test/test_os.py

```diff
@@ -2873,6 +2874,90 @@ def test_getfinalpathname_handles(self):
 
         self.assertEqual(0, handle_delta)
 
+    @support.requires_subprocess()
+    def test_stat_unlink_race(self):
+        # bpo-46785: the implementation of os.stat() falls back to reading
+        # the parent directory if CreateFileW() fails with a permission
+        # error. If reading the parent directory fails because the file or
+        # directory are subsequently unlinked, or because the volume or
+        # share are no longer available, then the original permission error
+        # should not be restored.
+        filename =  os_helper.TESTFN
+        self.addCleanup(os_helper.unlink, filename)
+        deadline = time.time() + 5
+        command = textwrap.dedent("""\
+            import os
+            import sys
+            import time
+
+            filename = sys.argv[1]
+            deadline = float(sys.argv[2])
+
+            while time.time() < deadline:
+                try:
+                    with open(filename, "w") as f:
+                        pass
+                except OSError:
+                    pass
+                try:
+                    os.remove(filename)
+                except OSError:
+                    pass
+            """)
+
+        with subprocess.Popen([sys.executable, '-c', command, filename, str(deadline)]) as proc:
+            while time.time() < deadline:
+                try:
+                    os.stat(filename)
+                except FileNotFoundError as e:
+                    assert e.winerror == 2  # ERROR_FILE_NOT_FOUND
+            try:
+                proc.wait(1)
+            except subprocess.TimeoutExpired:
+                proc.terminate()
+
+    @support.requires_subprocess()
+    def test_stat_rmdir_race(self):
```

Comment:
I don't think another test of `FileNotFoundError` is needed. Maybe there could be a test for the `ERROR_NOT_READY` case due to media removal, but I don't see how to set that up with the tools at hand, and it's so rare that it's not a practical problem. The far more likely case is that `CreateFileW()` fails with `ERROR_NOT_READY`, which isn't handled in `os.stat()`.

Reply:
- Regarding the test - why not? I don't like the code duplication, but the test isn't too complicated, and will verify in the future that no regressions occur. 
- Regarding `ERROR_NOT_READY` not being handled - I didn't notice that. It seems to me like handling `ERROR_NOT_READY` and `ERROR_BAD_NET_NAME` in this PR doesn't make much sense then, and should be done (and tested manually/programmatically) in another PR. (By the way, is there an issue in the issue tracker for this?)

Reply:
The `ERROR_PATH_NOT_FOUND` case is basically the same condition as `ERROR_FILE_NOT_FOUND`. When `CreateFileW()` is called, the file and its parent directory still exist, but by the time `FindFirstFileW()` is called, one or both is gone. I can't see anyone ever modify the `os.stat()` implementation to remove handling of one error but not the other. These two errors are the far more likely cases, e.g. some other thread is in removing the tree with something like `shutil.rmtree()`.

`ERROR_PATH_NOT_FOUND` would also occur if a removable drive gets removed with freakish coincidental timing.  That's basically the same case as `ERROR_BAD_NETNAME` (share removed) and `ERROR_NOT_READY` (removable media removed). These are all very unlikely in practice, but they're possible, and they fall into the same general category of the file no longer exists, and returning a `PermissionError` for ERROR_ACCESS_DENIED (5) or `ERROR_SHARING_VIOLATION` (32) for a file that no longer exists is pointless. It's hard to test these cases, but that doesn't mean they should be excluded from the PR entirely.

Reply:
> The `ERROR_PATH_NOT_FOUND` case is basically the same condition as `ERROR_FILE_NOT_FOUND`.

I don't disagree. I'll remove the test.

> It's hard to test these cases, but that doesn't mean they should be excluded from the PR entirely.

I'm fine with including them, but I also wanted to make sure that there's an issue for the more general case you mentioned.